draft-rescorla-tls-ctls.md

```diff
@@ -121,38 +121,62 @@ with a varint.
+The only cTLS records that are sent in plaintext are handshake records
+(ClientHello, and ServerHello/HRR). The content type is therefore constant (it
+is always `handshake`) and does not need to be explicitly conveyed in the record
+layer. Length information is signaled for unencrypted records using the varint
+encoding. (In principle, even this length could be omitted, since handshake
+messages are self-describing. Having this length field allows a clean separation
+between the record layer and the handshake layer.)
 
 ~~~~
       struct {
-          ContentType type;
-          opaque fragment[TLSPlaintext.length];
+          opaque fragment<0..V>;
       } TLSPlaintext;
 ~~~~
 
-In addition, because the epoch is known in advance, the
-dummy content type is not needed for the ciphertext, so
-TLSCiphertext becomes:
+Encrypted records use DTLS 1.3 record framing, comprising a configuration octet
+followed by optional connection ID, sequence number, and length fields.    
 
 ~~~~
-      struct {
-          opaque content[TLSPlaintext.length];
-          ContentType type;
-          uint8 zeros[length_of_padding];
-      } TLSInnerPlaintext;
+      0 1 2 3 4 5 6 7
+      +-+-+-+-+-+-+-+-+
+      |0|0|1|C|S|L|E E|
+      +-+-+-+-+-+-+-+-+
+      | Connection ID |   Legend:
+      | (if any,      |
+      /  length as    /   C   - Connection ID (CID) present
+      |  negotiated)  |   S   - Sequence number length
+      +-+-+-+-+-+-+-+-+   L   - Length present
+      | 8 or 16 bit   |   E   - Epoch
+      |Sequence Number|
+      | (if present)  |
+      +-+-+-+-+-+-+-+-+
+      | 16 bit Length |
+      | (if present)  |
+      +-+-+-+-+-+-+-+-+
 
       struct {
-          opaque encrypted_record[TLSCiphertext.length];
-      } TLSCiphertext;
+          opaque unified_hdr[variable];
+          opaque encrypted_record[length];
+      } DTLSCiphertext;
 ~~~~
 
-Note: The user is responsible for ensuring that the sequence
-numbers/nonces are handled in the usual fashion.
+The presence and size of the connection ID field is negotiated as in DTLS.
+
+As with DTLS, the length field MAY be omitted by clearing the L bit, which means
+that the record consumes the entire rest of the data in the lower level
+transport.  In this case it is not possible to have multiple DTLSCiphertext
+format records without length fields in the same datagram.  In stream-oriented
+transports (e.g., TCP), the length field MUST be present. For use over other 
+transports length information may be inferred from the underlying layer. 
 
+Normal DTLS does not provide a mechanism for suppressing the sequence number
+field entirely.  In cases where a sequence number is not required (e.g., when a
+reliable transport is in use), a cTLS implementation may suppress it by setting
+the `suppressSequenceNumber` flag in the compression profile being used (see
+{{specifying-a-specialization}}).  When this flag is enabled, the S bit in the
+configuration octet MUST be cleared.
```

Comment:
This is just to have a fixed encoding, right? As it would ordinarily indicate an 8-bit length.

Reply:
Correct.  Either value is wrong (since there's no length field), so we might as well choose one.